### PR TITLE
Extend structures and structure pools mappings

### LINKED ISSUE #1129: Structure names
@falseresync closed https://github.com/FabricMC/yarn/pull/1115, but there were lots of good names in it. Making an issue so that we don't forget about them and add them in a future PR.

### PR DESCRIPTION
**Pull request by falseresync**
Closes #1129

### CODE REVIEW

File: mappings/net/minecraft/structure/pool/StructurePoolElement.mapping [hunk 1]
```diff
@@ -3,6 +3,7 @@ CLASS net/minecraft/class_3784 net/minecraft/structure/pool/StructurePoolElement
 	METHOD method_16622 setProjection (Lnet/minecraft/class_3785$class_3786;)Lnet/minecraft/class_3784;
 		ARG 1 projection
 	METHOD method_16624 getProjection ()Lnet/minecraft/class_3785$class_3786;
+	METHOD method_16625 toRawDynamic (Lcom/mojang/datafixers/types/DynamicOps;)Lcom/mojang/datafixers/Dynamic;
```
Comment (Runemoro) [inline mappings/net/minecraft/structure/pool/StructurePoolElement.mapping]: What does "raw" mean?

Comment (falseresync) [reply mappings/net/minecraft/structure/pool/StructurePoolElement.mapping]: There’re 2 `toDynamic` methods: one is protected and implemented in children, the other one is public, implemented in the parent and calls the protected one. As they have the identical signatures it is necessary to give them different names, so I used `Raw` to indicate that this is not a final dynamic, which the consumer will get

Comment (Runemoro) [reply mappings/net/minecraft/structure/pool/StructurePoolElement.mapping]: But what do they do differently? Does the public one just call the protected one, or does it also do something more?

Comment (falseresync) [reply mappings/net/minecraft/structure/pool/StructurePoolElement.mapping]: The public one merges some additional info with the result of the protected one

Comment (liach) [reply mappings/net/minecraft/structure/pool/StructurePoolElement.mapping]: Hmm, so like `contentToDynamic` instead?

File: mappings/net/minecraft/structure/Structure.mapping [hunk 2]
```diff
@@ -6,11 +6,31 @@ CLASS net/minecraft/class_3499 net/minecraft/structure/Structure
 	METHOD method_15160 getSize ()Lnet/minecraft/class_2338;
 	METHOD method_15161 setAuthor (Ljava/lang/String;)V
 		ARG 1 name
+	METHOD method_15162 addRotatedAndMirroredOffset (Lnet/minecraft/class_2338;Lnet/minecraft/class_2415;Lnet/minecraft/class_2470;II)Lnet/minecraft/class_2338;
```
Comment (liach) [inline mappings/net/minecraft/structure/Structure.mapping]: maybe `map` or `transform`, and arg3 and arg 4 are like `sizeX` and `sizeZ`

File: mappings/net/minecraft/structure/Structure.mapping [hunk 3]
```diff
@@ -6,11 +6,31 @@ CLASS net/minecraft/class_3499 net/minecraft/structure/Structure
...
 	METHOD method_15164 addEntitiesFromWorld (Lnet/minecraft/class_1937;Lnet/minecraft/class_2338;Lnet/minecraft/class_2338;)V
 		ARG 1 world
 		ARG 2 firstCorner
 		ARG 3 secondCorner
+	METHOD method_15165 getInfosForBlock (Lnet/minecraft/class_2338;Lnet/minecraft/class_3492;Lnet/minecraft/class_2248;Z)Ljava/util/List;
+		ARG 1 pos
+		ARG 2 placementData
+		ARG 4 mirroredAndRotated
+	METHOD method_15166 getRotatedSize (Lnet/minecraft/class_2470;)Lnet/minecraft/class_2338;
+	METHOD method_15167 addRotatedAndMirroredSize (Lnet/minecraft/class_2338;Lnet/minecraft/class_2415;Lnet/minecraft/class_2470;)Lnet/minecraft/class_2338;
```
Comment (liach) [inline mappings/net/minecraft/structure/Structure.mapping]: `transform` or `map` (a block pos)

File: mappings/net/minecraft/structure/Structure.mapping [hunk 4]
```diff
@@ -6,11 +6,31 @@ CLASS net/minecraft/class_3499 net/minecraft/structure/Structure
...
+		ARG 3 offsetX
+		ARG 4 offsetZ
 	METHOD method_15164 addEntitiesFromWorld (Lnet/minecraft/class_1937;Lnet/minecraft/class_2338;Lnet/minecraft/class_2338;)V
 		ARG 1 world
 		ARG 2 firstCorner
 		ARG 3 secondCorner
+	METHOD method_15165 getInfosForBlock (Lnet/minecraft/class_2338;Lnet/minecraft/class_3492;Lnet/minecraft/class_2248;Z)Ljava/util/List;
+		ARG 1 pos
+		ARG 2 placementData
+		ARG 4 mirroredAndRotated
```
Comment (liach) [inline mappings/net/minecraft/structure/Structure.mapping]:
```suggestion
		ARG 4 transformed
```

File: mappings/net/minecraft/structure/Structure.mapping [hunk 5]
```diff
@@ -6,11 +6,31 @@ CLASS net/minecraft/class_3499 net/minecraft/structure/Structure
...
+	METHOD method_15166 getRotatedSize (Lnet/minecraft/class_2470;)Lnet/minecraft/class_2338;
+	METHOD method_15167 addRotatedAndMirroredSize (Lnet/minecraft/class_2338;Lnet/minecraft/class_2415;Lnet/minecraft/class_2470;)Lnet/minecraft/class_2338;
+	METHOD method_15168 mirrorAndRotateAroundPos (Lnet/minecraft/class_2338;Lnet/minecraft/class_2415;Lnet/minecraft/class_2470;Lnet/minecraft/class_2338;)Lnet/minecraft/class_2338;
```
Comment (liach) [inline mappings/net/minecraft/structure/Structure.mapping]: Maybe `mapAround` `transformAround`

File: mappings/net/minecraft/structure/Structure.mapping [hunk 6]
```diff
@@ -6,11 +6,31 @@ CLASS net/minecraft/class_3499 net/minecraft/structure/Structure
...
+		ARG 0 pos
+		ARG 3 rotationCenter
```
Comment (liach) [inline mappings/net/minecraft/structure/Structure.mapping]:
```suggestion
		ARG 3 pivot
```

File: mappings/net/minecraft/structure/Structure.mapping [hunk 7]
```diff
@@ -6,11 +6,31 @@ CLASS net/minecraft/class_3499 net/minecraft/structure/Structure
...
 	METHOD method_15169 createIntListTag ([I)Lnet/minecraft/class_2499;
+	METHOD method_15171 mirrorAndRotate (Lnet/minecraft/class_3492;Lnet/minecraft/class_2338;)Lnet/minecraft/class_2338;
```
Comment (liach) [inline mappings/net/minecraft/structure/Structure.mapping]: `mapBy` `transformBy`

File: mappings/net/minecraft/structure/pool/FeaturePoolElement.mapping [hunk 8]
```diff
@@ -3,3 +3,5 @@ CLASS net/minecraft/class_3776 net/minecraft/structure/pool/FeaturePoolElement
 	FIELD field_16662 tag Lnet/minecraft/class_2487;
 	METHOD <init> (Lnet/minecraft/class_2975;)V
 		ARG 1 feature
+	METHOD method_16601 getStart (Lnet/minecraft/class_3485;Lnet/minecraft/class_2470;)Lnet/minecraft/class_2338;
+	METHOD method_19299 getDefaultJigsawTag ()Lnet/minecraft/class_2487;
```
Comment (liach) [inline mappings/net/minecraft/structure/pool/FeaturePoolElement.mapping]:
```suggestion
	METHOD method_19299 createDefaultJigsawTag ()Lnet/minecraft/class_2487;
```

**Review by liach — CHANGES_REQUESTED** [review]
Quite a few problems... "mirror and rotate" can just be replaced by `transform`/`map` wholesale imo, and this is just a preliminary review (still more to go into depth)

File: mappings/net/minecraft/structure/pool/StructurePoolElement.mapping (hunk 1 above)
Comment (falseresync) [reply mappings/net/minecraft/structure/pool/StructurePoolElement.mapping]: I don’t see much difference

File: mappings/net/minecraft/structure/Structure.mapping (hunk 2 above)
Comment (falseresync) [reply mappings/net/minecraft/structure/Structure.mapping]: `map`/`transform` sound a bit too generic to my taste, maybe someone else also has a thought on that

File: mappings/net/minecraft/structure/pool/StructurePoolElement.mapping (hunk 1 above)
Comment (liach) [reply mappings/net/minecraft/structure/pool/StructurePoolElement.mapping]: Then should be `rawToDynamic` than `toRawDynamic`

Comment (falseresync) [reply mappings/net/minecraft/structure/pool/StructurePoolElement.mapping]: True

File: mappings/net/minecraft/structure/Structure.mapping (hunk 2 above)
Comment (liach) [reply mappings/net/minecraft/structure/Structure.mapping]: "mirror and rotate" is way too wordy. A structure can have a position remapped by a mirror and a rotation combination, so `map` or `remap` is accurate.

Comment (liach) [reply mappings/net/minecraft/structure/Structure.mapping]: @Runemoro want to hear from you here.

Comment (Runemoro) [reply mappings/net/minecraft/structure/Structure.mapping]: The operation of mirroring, rotating, and offsetting is called an [isometry](https://en.m.wikipedia.org/wiki/Isometry). But I don't know how to turn that into a method name... `applyIsometry`? `isometricTransform`? `transform`?

Comment (falseresync) [reply mappings/net/minecraft/structure/Structure.mapping]: I might be wrong, but the actual code is about mirroring and rotating the given _offset_, and then applying it to the position, which is even trickier

Comment (falseresync) [reply mappings/net/minecraft/structure/Structure.mapping]: I might be wrong, but the actual code is about mirroring and rotating the given _offset_, and then applying it to the position, which is even trickier

Comment (Runemoro) [reply mappings/net/minecraft/structure/Structure.mapping]: Then `applyTransformedOffset`. The specific transformations applied to the offset should just be in the parameter names.

Comment (falseresync) [reply mappings/net/minecraft/structure/Structure.mapping]: What’s about its derivative `addMirroredAndRotatedSize`, which uses structure size as offset; is it good as `offsetByTransformedSize`?

Comment (liach) [reply mappings/net/minecraft/structure/Structure.mapping]: Should rename that `offset ` to `pivot`

Comment (falseresync) [reply mappings/net/minecraft/structure/Structure.mapping]: In this case it is an offset, not a center of the rotation

Comment (liach) [reply mappings/net/minecraft/structure/Structure.mapping]: Hmm, `applyOffsetTransform`?

Comment (falseresync) [reply mappings/net/minecraft/structure/Structure.mapping]: It applies transformed offset, not offset(ted) transformation

Comment (liach) [reply mappings/net/minecraft/structure/Structure.mapping]: Then I support runemoro's `applyTransformedOffset`

Comment (falseresync) [reply mappings/net/minecraft/structure/Structure.mapping]: Already committed that

File: mappings/net/minecraft/structure/pool/StructurePool.mapping [hunk 9]
```diff
@@ -18,6 +19,7 @@ CLASS net/minecraft/class_3785 net/minecraft/structure/pool/StructurePool
 	METHOD method_16633 getElementIndicesInRandomOrder (Ljava/util/Random;)Ljava/util/List;
 		ARG 1 random
 	METHOD method_16634 getTerminatorsId ()Lnet/minecraft/class_2960;
+	METHOD method_19309 getLargestY (Lnet/minecraft/class_3485;)I
```
Comment (liach) [inline mappings/net/minecraft/structure/pool/StructurePool.mapping]: Think we usually say "highest"

**Review by liach — APPROVED** [review]

File: mappings/net/minecraft/structure/pool/StructurePool.mapping (hunk 9 above)
Comment (falseresync) [reply mappings/net/minecraft/structure/pool/StructurePool.mapping]: Fixed

File: mappings/net/minecraft/structure/pool/StructurePool.mapping [hunk 10]
```diff
@@ -6,6 +6,7 @@ CLASS net/minecraft/class_3785 net/minecraft/structure/pool/StructurePool
 	FIELD field_16746 INVALID Lnet/minecraft/class_3785;
 	FIELD field_16863 projection Lnet/minecraft/class_3785$class_3786;
 	FIELD field_16864 elementCounts Lcom/google/common/collect/ImmutableList;
+	FIELD field_18707 largestY I
```
Comment (liach) [inline mappings/net/minecraft/structure/pool/StructurePool.mapping]:
```suggestion
	FIELD field_18707 highestY I
```

**Review by jaskarth — APPROVED** [review]
good pr

**Review by liach — APPROVED** [review]

**Review by Juuxel — APPROVED** [review]

**Review by Runemoro — APPROVED** [review]
